JSIL/SpecialIdentifiers.cs

```diff
@@ -111,18 +111,43 @@ JSSpecialIdentifiers js
             return new JSNewArrayExpression(elementType, dimensions, initializer);
         }
 
-        public JSInvocationExpression NewDelegate (TypeReference delegateType, JSExpression thisReference, JSExpression targetMethod, JSMethod method) {
+        public JSInvocationExpression NewDelegate (TypeReference delegateType, JSExpression thisReference, JSExpression targetMethod)
+        {
+            var targetDotExpression = targetMethod as JSDotExpressionBase;
+            var jsMethod = targetDotExpression != null ? targetDotExpression.Member as JSMethod : null;
+            var jsMethodAccess = targetMethod as JSMethodAccess;
+
+            var arguments = jsMethod == null ? null : jsMethod.Reference
+                .Parameters.Select(
+                    (parameter, index) =>
+                        (JSExpression) new JSRawOutputIdentifier(parameter.ParameterType, "arguments[{0}]", index))
+                .ToArray();
+
+            bool isDelegateCreation = jsMethod != null && jsMethod.Reference.Name == "Invoke" &&
+                                      TypeUtil.IsDelegateType(jsMethod.Reference.DeclaringType);
+
             return JSInvocationExpression.InvokeStatic(
                 new JSDotExpression(
                     new JSType(delegateType),
-                    new JSFakeMethod("New", delegateType, new[] { TypeSystem.Object, TypeSystem.Object }, MethodTypes)
-                ), 
-                method == null ? new [] { thisReference, targetMethod } : new [] {
-                    thisReference, 
-                    targetMethod, 
-                    new JSDefferedExpression(new JSMethodOfExpression(method.Reference, method.Method, method.MethodTypes, method.GenericArguments)),  },
+                    new JSFakeMethod("New", delegateType, new[] {TypeSystem.Object, TypeSystem.Object}, MethodTypes)
+                    ),
+                jsMethod == null
```

Comment:
Can you split this out into some named locals? This is way too complicated for a single statement.